File: apps/ios-wiki-demo/App/Model/ChangeMark.swift
import Foundation
import Xtriever

/// How a hit's position changed between the fused list and the re-ranked list (spec US2;
/// data-model "ChangeMark"). A pure function of two ranked lists, keyed by external id.
enum ChangeMark: Equatable {
    /// In the re-ranked list only.
    case new
    /// Same rank in both.
    case same
    /// Rose by `n` positions.
    case up(Int)
    /// Fell by `n` positions.
    case down(Int)

    /// Marks for every re-ranked hit, and the fused hits that fell out of the re-ranked list.
    static func compute(fused: [Hit], reranked: [Hit]) -> (marks: [String: ChangeMark], dropped: [Hit]) {
        var fusedRank: [String: Int] = [:]
        for (rank, hit) in fused.enumerated() { fusedRank[hit.externalId] = rank }
        var marks: [String: ChangeMark] = [:]
        var seen: Set<String> = []
        for (rank, hit) in reranked.enumerated() {
            seen.insert(hit.externalId)
            guard let before = fusedRank[hit.externalId] else { marks[hit.externalId] = .new; continue }
            let delta = before - rank
            marks[hit.externalId] = delta == 0 ? .same : (delta > 0 ? .up(delta) : .down(-delta))
        }
        let dropped = fused.filter { !seen.contains($0.externalId) }
        return (marks, dropped)
    }
}

File: apps/android-wiki-demo/src/main/kotlin/dev/xtriever/demo/ChangeMark.kt
package dev.xtriever.demo

/**
 * How a hit's position changed between the fused list and the re-ranked one (Feature 025,
 * spec FR-006). A pure function of two ranked lists keyed by external identifier — the rule
 * the iOS demo states in `ChangeMark.swift`, ported so both demonstrations mark the same
 * movement the same way.
 *
 * This is the only arithmetic the application performs. Everything else on screen is the
 * engine's own number.
 */
sealed interface ChangeMark {
    /** In the re-ranked list only. */
    data object New : ChangeMark

    /** Same rank in both lists. */
    data object Unchanged : ChangeMark

    /** Rose by [places] positions. */
    data class Up(val places: Int) : ChangeMark

    /** Fell by [places] positions. */
    data class Down(val places: Int) : ChangeMark

    companion object {
        /**
         * Marks for every re-ranked hit, and the identifiers of the fused hits that fell out of
         * the re-ranked head.
         */
        fun compute(fused: List<String>, reranked: List<String>): Pair<Map<String, ChangeMark>, List<String>> {
            val fusedRank = fused.withIndex().associate { (rank, id) -> id to rank }
            val marks = reranked.withIndex().associate { (rank, id) ->
                val before = fusedRank[id]
                id to when {
                    before == null -> New
                    before == rank -> Unchanged
                    before > rank -> Up(before - rank)
                    else -> Down(rank - before)
                }
            }
            val kept = reranked.toSet()
            return marks to fused.filterNot { it in kept }
        }
    }
}
